[[ Bug 18443 ]] Reset private clipboard type when clipboard modified

runrevmark opened the pull request:
When another application modifies the clipboard, the private clipboard
data must be flushed.

runrevmark committed:
[[ Bug 18443 ]] Reset private clipboard type when clipboard modified

When another application modifies the clipboard, the private clipboard
data must be flushed.

runrevmark committed:
[[ Bug 18443 ]] Add test for private clipboard resetting

This patch adds a test for private clipboard resetting. In addition
it fixes the resource references in the other clipboard tests (the
_Tiny*() functions) and adds a 'TestRunStack' library command which
executes a stack using the engine being tested as a separate
process.

livecodeali commented on tests/_testlib.livecodescript:
```diff
@@ -306,27 +306,41 @@ on TestRepeat pDesc, pHandler, pTarget, pTimeOut, pParamsArray
+	put format("\"%s\" %s \"%s\"", tEnginePath, pOptions, pStackFilePath) into tCommand
+
+	local tOutput
+	put shell(tCommand) into tOutput
```
Either here or in the test you'll need to make sure that `-ui` is in the options if "command line" is in the enviroment, otherwise this will fail on Travis, I think (similar to the problem I've just had with standalone builder tests!)

runrevmark replied:
I think the clipboard tests are skipped in -ui mode - so I don't think they are called at all?

runrevmark replied:
(i.e. it makes no sense to pass '-ui' here, as if the clipboard doesn't work in -ui mode, neither is the subprocess's function).

runrevmark committed:
[[ Bug 18443 ]] Make skipping of clipboard tests more accurate

The clipboard tests can be run in either -ui or normal mode on
Mac and Windows.

This patch ensures that the tests are only skipped on Linux if
the tests are run in -ui mode.

livecodeali commented:
@livecode-vulcan review ok d310b0a

runrevmark commented:
@livecodefraser : I've had to make the private data member of MCClipboard 'mutable' here due to much constness. If you could confirm this is okay before merging that would be great - if not, then const-ness will need to be tweaked in a fair few methods.

livecode-vulcan commented:
💙 review by @livecodeali ok d310b0af4b8627f583f07decc2eb54c5b57df207

livecode-vulcan commented:
😎 test success d310b0af4b8627f583f07decc2eb54c5b57df207

* try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/793)
* try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/798)
* try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/791)
* try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/792)
* try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/798)
* try-community-win-x86: [success](https://vulcan.livecode.com/builders/try-community-win-x86/builds/790)
* try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/790)

runrevmark commented:
Superceeded by #5060
